### Problem Statement :- IMDB and Amazon Review Classification :

#### Solution :- IMDB and Amazon Review Classification with Spacy :

#### Import Libraries :

In [157]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

#### Import Dataset :

In [158]:
data_yelp = pd.read_csv(r'C:\Users\PALLAVI\Downloads\yelp_labelled.txt', sep='\t',header=None)

#### View Top 5 Rows :

In [159]:
data_yelp.head()

0  1
0                           Wow... Loved this place.  1
1                                 Crust is not good.  0
2          Not tasty and the texture was just nasty.  0
3  Stopped by during the late May bank holiday of...  1
4  The selection on the menu was great and so wer...  1

 review and sentiment
 0-Negative, 1-Positive for positive review

#### View Bottom 5 Rows :

In [160]:
data_yelp.tail()

0  1
995  I think food should have flavor and texture an...  0
996                           Appetite instantly gone.  0
997  Overall I was not impressed and would not go b...  0
998  The whole experience was underwhelming, and I ...  0
999  Then, as if I hadn't wasted enough of my life ...  0

### Assign column names :


In [161]:
columan_name = ['Review', 'Sentiment']
data_yelp.columns = columan_name

In [162]:
data_yelp.head()

Review  Sentiment
0                           Wow... Loved this place.          1
1                                 Crust is not good.          0
2          Not tasty and the texture was just nasty.          0
3  Stopped by during the late May bank holiday of...          1
4  The selection on the menu was great and so wer...          1

In [163]:
data_yelp.shape


(1000, 2)

1000 rows (reviews), 2 columns (Sentiments)

#### Import Dataset :

In [164]:
data_amazon = pd.read_csv(r'C:\Users\PALLAVI\Downloads\amazon_cells_labelled.txt',sep='\t',header=None)


#### View Top 5 Rows :

In [165]:
data_amazon.head()

0  1
0  So there is no way for me to plug it in here i...  0
1                        Good case, Excellent value.  1
2                             Great for the jawbone.  1
3  Tied to charger for conversations lasting more...  0
4                                  The mic is great.  1

 review and sentiment
 0-Negative, 1-Positive for positive review

#### View Bottom 5 Rows :

In [166]:
data_amazon.tail()

0  1
995  The screen does get smudged easily because it ...  0
996  What a piece of junk.. I lose more calls on th...  0
997                       Item Does Not Match Picture.  0
998  The only thing that disappoint me is the infra...  0
999  You can not answer calls with the unit, never ...  0

In [167]:
data_amazon.columns = columan_name

In [168]:
data_amazon.head()

Review  Sentiment
0  So there is no way for me to plug it in here i...          0
1                        Good case, Excellent value.          1
2                             Great for the jawbone.          1
3  Tied to charger for conversations lasting more...          0
4                                  The mic is great.          1

In [169]:
data_amazon.shape

(1000, 2)

#### Import Dataset :

In [170]:
data_imdb = pd.read_csv(r'C:\Users\PALLAVI\Downloads\imdb_labelled.txt',sep='\t',header=None)

#### View Top 5 Rows :

In [171]:
data_imdb.head()

0  1
0  A very, very, very slow-moving, aimless movie ...  0
1  Not sure who was more lost - the flat characte...  0
2  Attempting artiness with black & white and cle...  0
3       Very little music or anything to speak of.    0
4  The best scene in the movie was when Gerardo i...  1

#### View Bottom 5 Rows :

In [172]:
data_imdb.tail()

0  1
743  I just got bored watching Jessice Lange take h...  0
744  Unfortunately, any virtue in this film's produ...  0
745                   In a word, it is embarrassing.    0
746                               Exceptionally bad!    0
747  All in all its an insult to one's intelligence...  0

In [173]:
data_imdb.columns = columan_name

In [174]:
data_imdb.head()

Review  Sentiment
0  A very, very, very slow-moving, aimless movie ...          0
1  Not sure who was more lost - the flat characte...          0
2  Attempting artiness with black & white and cle...          0
3       Very little music or anything to speak of.            0
4  The best scene in the movie was when Gerardo i...          1

In [175]:
data_imdb.shape

(748, 2)

#### Append all the data in a single DataFrame :

In [176]:
data = data_yelp.append([data_amazon, data_imdb],ignore_index=True)

In [177]:
data.shape

(2748, 2)

In [178]:
data.head()

Review  Sentiment
0                           Wow... Loved this place.          1
1                                 Crust is not good.          0
2          Not tasty and the texture was just nasty.          0
3  Stopped by during the late May bank holiday of...          1
4  The selection on the menu was great and so wer...          1

#### Check Distribution of Sentiments  : 

In [179]:
data['Sentiment'].value_counts()

1    1386
0    1362
Name: Sentiment, dtype: int64

1386 positive reviews
1362 Negative reviews

#### Check for Null values :


In [180]:
data.isnull().sum()

Review       0
Sentiment    0
dtype: int64

#### Split "data" into Independent & Dependent Variable :

In [181]:
x = data['Review']
y = data['Sentiment']

#### Data Cleaning :


In [182]:
import string

In [183]:
punct = string.punctuation

In [184]:
punct

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'

In [185]:
from spacy.lang.en.stop_words import STOP_WORDS

In [186]:
stopwords = list(STOP_WORDS) # list of stopwords

In [187]:
stopwords

['ever',
 'therein',
 'forty',
 'may',
 'well',
 'anywhere',
 'keep',
 'three',
 'thru',
 'whole',
 'hence',
 'him',
 'becoming',
 'n‘t',
 'nine',
 '‘ll',
 'of',
 'over',
 'some',
 'while',
 'regarding',
 'seemed',
 '’s',
 'really',
 'thereafter',
 'almost',
 'otherwise',
 'himself',
 'his',
 'mostly',
 'nobody',
 're',
 '’m',
 'it',
 'afterwards',
 'show',
 'although',
 'whereas',
 'around',
 'were',
 'under',
 'why',
 'whose',
 'this',
 'become',
 'front',
 'using',
 'into',
 'its',
 'everywhere',
 'nevertheless',
 "'ll",
 'indeed',
 "'s",
 'or',
 'hereupon',
 'serious',
 'can',
 'others',
 'within',
 'much',
 'twenty',
 'might',
 'four',
 'not',
 'go',
 'through',
 'all',
 'their',
 'behind',
 'someone',
 'many',
 'call',
 'along',
 '‘ve',
 'whither',
 'together',
 'further',
 'have',
 'onto',
 'but',
 "'m",
 'due',
 'who',
 'between',
 'seem',
 'hers',
 'just',
 'always',
 'moreover',
 'put',
 'below',
 'do',
 'against',
 'anyone',
 'as',
 'itself',
 'upon',
 'you',
 'every',
 'you

#### Creat a function for data cleaning :

In [188]:
import spacy 

nlp = spacy.load('en_core_web_sm')

def text_data_cleaning(sentence):
    doc = nlp(sentence)

    tokens = [] # list of tokens
    for token in doc:
        if token.lemma_ != "-PRON-":
            temp = token.lemma_.lower().strip()
        else:
            temp = token.lower_
        tokens.append(temp)
  
    cleaned_tokens = []
    for token in tokens:
        if token not in stopwords and token not in punct:
            cleaned_tokens.append(token)
    return cleaned_tokens

 If root form of that word is not pronoun then it is going to convert that into lower form
 and if that word is a proper noun, then we are directly taking lower form, because there is no lemma for proper noun

#### stopwords and punctuations removed :

In [189]:
text_data_cleaning("Hello all, It's a beautiful day outside there!")


['hello', 'beautiful', 'day', 'outside']

#### Vectorization Feature Engineering (TF-IDF) :


In [190]:
from sklearn.svm import LinearSVC
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.pipeline import Pipeline

In [191]:
tfidf = TfidfVectorizer(tokenizer=text_data_cleaning)

tokenizer = text_data_cleaning, tokenization will be done according to this function.

In [192]:
classifier = LinearSVC()

### Train the model :

1.   List item
2.   List item



#### Splitting the dataset into the Train and Test set :

In [193]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 0)

In [194]:
x_train.shape, x_test.shape

((2198,), (550,))

In [195]:
x_train.head()

2572    An Italian reviewer called this "a small, grea...
526                          And it was way to expensive.
1509    As an earlier review noted, plug in this charg...
144     Nice blanket of moz over top but i feel like t...
2483    The film gives meaning to the phrase, "Never i...
Name: Review, dtype: object

#### Fit the x_train and y_train :

In [196]:
clf = Pipeline([('tfidf',tfidf), ('clf',classifier)])

It will first do vectorization and then it will do classification .

In [197]:
clf.fit(x_train, y_train)

Pipeline(steps=[('tfidf',
                 TfidfVectorizer(tokenizer=<function text_data_cleaning at 0x000002286B0A5E50>)),
                ('clf', LinearSVC())])

In this we don't need to prepare the dataset for testing(x_test).

### Predict the Test set results :

*   List item
*   List item



In [198]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

In [199]:
y_pred = clf.predict(x_test)

In [200]:
y_pred

array([1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0,
       0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0,
       1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1,
       0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 0,
       0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1,
       1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0,
       1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0,
       0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0,
       1, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0,

### Confusion matrix :


In [201]:
confusion_matrix(y_test, y_pred)

array([[204,  75],
       [ 49, 222]], dtype=int64)

### classification report :


In [202]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.81      0.73      0.77       279
           1       0.75      0.82      0.78       271

    accuracy                           0.77       550
   macro avg       0.78      0.78      0.77       550
weighted avg       0.78      0.77      0.77       550



#### Accuracy : 

In [203]:
accuracy_score(y_test, y_pred)


0.7745454545454545

#### Now, Test our model with "New Data" :

In [204]:
clf.predict(["Wow, I am learning Natural Language Processing in fun fashion!"])


array([1], dtype=int64)

output is 1, that means review is Positive.

In [205]:
clf.predict(["It's hard to learn new things!"])

array([0], dtype=int64)

 output is 0, that means review is Negative.

#### ................................... Thank You .....................................................................................